### PR TITLE
[EVM] Use Keccak256 hash-ed linker symbols in LLVM MC/LLD

### PR DESCRIPTION
**Pull request by PavelKopyl**
The patch is required to overcome the issue with nasty symbols in the datasize/dataoffset arguments. Instead of using file paths as symbols names we use their Keccak256 hashes.   

# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp [hunk 1]
```diff
@@ -105,6 +125,49 @@ bool EVMAsmPrinter::isBlockOnlyReachableByFallthrough(
...
+  StringRef LinkerSymbolName = LinkerSymbol->getName();
+  unsigned Opc = MI->getOpcode();
+
+  std::string SymbolNameHash = EVM::getLinkerSymbolHash(LinkerSymbolName);
+  std::string DataSymbolNameHash =
+      (Opc == EVM::DATASIZE_S) ? EVM::getDataSizeSymbol(SymbolNameHash)
+                               : EVM::getDataOffsetSymbol(SymbolNameHash);
+
+  MCInst MCI;
+  if (Opc == EVM::DATASIZE_S || Opc == EVM::DATAOFFSET_S)
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: Can we add an assert in the beginning of the function that these two opcodes are expected? If that is possible, we don't need this check here.

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp [hunk 2]
```diff
@@ -105,6 +125,49 @@ bool EVMAsmPrinter::isBlockOnlyReachableByFallthrough(
...
+
+  // The linker symbol and the related section already exist, so just exit.
+  if (LinkerSymbolSet.contains(SymbolNameHash))
+    return;
+
+  LinkerSymbolSet.insert(SymbolNameHash);
+
+  MCSection *CurrentSection = OutStreamer->getCurrentSectionOnly();
+
+  // Emit the .linker_symbol_name section that contains the actual symbol
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: Can we emit this section at the end of a file?
Currently this section is generated in the .text section:
```
foo:                                    ; @foo
; %bb.0:                                ; %entry
	JUMPDEST
	PUSH4 @__dataoffset___$0336fd807c0716a535e520df5b63ecc41ba7984875fdfa2241fcf3c8d0107e26$__@evm_data
	.linker_symbol_name__$0336fd807c0716a535e520df5b63ecc41ba7984875fdfa2241fcf3c8d0107e26$__
	.ascii	"D_105_deployed"
	.text
	PUSH4 @__datasize___$0336fd807c0716a535e520df5b63ecc41ba7984875fdfa2241fcf3c8d0107e26$__@evm_data
	ADD
	DUP2
	DUP2
	SWAP3
	POP
	SWAP1
	POP
	JUMP
```

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 3]
```diff
@@ -518,33 +522,40 @@ static std::string creteEVMLinkerScript(ArrayRef<LLVMMemoryBufferRef> memBufs,
...
-    symDataOffsetDeps = (dataOffsetPrefix + bufIDs[2] + " = .;\n").str();
+    symDataOffsetDeps =
+        EVM::getDataOffsetSymbol(EVM::getLinkerSymbolHash(bufIDs[2])) +
+        " = .;\n";
     for (unsigned idx = 3; idx < numObjectsToLink; ++idx)
       symDataOffsetDeps +=
-          (dataOffsetPrefix + bufIDs[idx] + " = " + dataOffsetPrefix +
-           bufIDs[idx - 1] + " + " + dataSizePrefix + bufIDs[idx - 1] + ";\n")
-              .str();
+          EVM::getDataOffsetSymbol(EVM::getLinkerSymbolHash(bufIDs[idx])) +
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: It's preferable to use `raw_string_ostream` to concatenate strings.
It's faster and will likely result in better readable code.

File: lld/lld-c/LLDAsLibraryC.cpp [hunk 4]
```diff
@@ -570,16 +581,21 @@ LLVMBool LLVMLinkEVM(LLVMMemoryBufferRef inBuffers[],
   assert(numInBuffers > 1);
   SmallVector<MemoryBufferRef> localInMemBufRefs(3);
   SmallVector<std::unique_ptr<MemoryBuffer>> localInMemBufs(3);
+
+  // TODO: verify that the object files contain sections with original
```
Comment (akiramenai) [inline lld/lld-c/LLDAsLibraryC.cpp]: Could you kindly create a ticket for that and mention it here?

Comment (vladimirradosavljevic) [inline lld/lld-c/LLDAsLibraryC.cpp]: If I understand correctly, we emit `.linker_symbol_name.` sections and place original name just to verify the names here? If that is correct, I would suggest not to emit these sections until we address this.

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp (hunk 1 above)
Comment (akiramenai) [reply llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: Indeed it seems that 
`DataSymbolNameHash` definition uses this assumption anyway.

Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: Thank you! I missed to remove this.

File: llvm/lib/Target/EVM/EVMAsmPrinter.cpp (hunk 2 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMAsmPrinter.cpp]: Sure, we can move it to the end. This just looks, as if we break `.text `section, but actually this syntax is fine. Here we just  switch from` .text ` to `.linker_symbol_name__$0336fd807c0716a535e520df5b63ecc41ba7984875fdfa2241fcf3c8d0107e26$__` section and then switch back to the ` .text `.
(Strictly speaking, the `.section` directive is missing before `.linker_symbol_name...`, because this is a custom section).
But I agree let's just remove it now and restore when we start using these sections.

File: lld/lld-c/LLDAsLibraryC.cpp (hunk 4 above)
Comment (PavelKopyl) [reply lld/lld-c/LLDAsLibraryC.cpp]: Right. Let's removed this.

**Review by vladimirradosavljevic — APPROVED** [review]
LGTM. Thanks!
Please also wait for @akiramenai to approve.

**Comment by vladimirradosavljevic** [comment]
@PavelKopyl Could you please squash commits to make CI green?

**Review by akiramenai — APPROVED** [review]
LGTM, please squash.

**Review by hedgar2017 — APPROVED** [review]
